File: apps/mobile/android/app/src/main/java/com/snowzlmbot/hermes/mobile/core/GatewayProtocol.kt
package com.snowzlmbot.hermes.mobile.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

private val protocolJson = Json {
  ignoreUnknownKeys = true
  encodeDefaults = false
}

class ProtocolException(message: String) : IllegalArgumentException(message)

enum class GatewayEventType {
  GATEWAY_READY,
  SESSION_INFO,
  SESSION_TITLE,
  MESSAGE_START,
  MESSAGE_DELTA,
  MESSAGE_INTERIM,
  MESSAGE_COMPLETE,
  REASONING_DELTA,
  THINKING_DELTA,
  STATUS_UPDATE,
  TOOL_START,
  TOOL_PROGRESS,
  TOOL_COMPLETE,
  APPROVAL_REQUEST,
  CLARIFY_REQUEST,
  CLARIFY_EXPIRE,
  SECRET_REQUEST,
  SECRET_EXPIRE,
  SUDO_REQUEST,
  SUDO_EXPIRE,
  ERROR,
  SESSIONS_CHANGED,
  UNKNOWN,
  ;

  companion object {
    fun fromWire(value: String): GatewayEventType = when (value) {
      "gateway.ready" -> GATEWAY_READY
      "session.info" -> SESSION_INFO
      "session.title" -> SESSION_TITLE
      "message.start" -> MESSAGE_START
      "message.delta" -> MESSAGE_DELTA
      "message.interim" -> MESSAGE_INTERIM
      "message.complete" -> MESSAGE_COMPLETE
      "reasoning.delta" -> REASONING_DELTA
      "thinking.delta" -> THINKING_DELTA
      "status.update" -> STATUS_UPDATE
      "tool.start" -> TOOL_START
      "tool.progress" -> TOOL_PROGRESS
      "tool.complete" -> TOOL_COMPLETE
      "approval.request" -> APPROVAL_REQUEST
      "clarify.request" -> CLARIFY_REQUEST
      "clarify.expire" -> CLARIFY_EXPIRE
      "secret.request" -> SECRET_REQUEST
      "secret.expire" -> SECRET_EXPIRE
      "sudo.request" -> SUDO_REQUEST
      "sudo.expire" -> SUDO_EXPIRE
      "error" -> ERROR
      "sessions.changed" -> SESSIONS_CHANGED
      else -> UNKNOWN
    }
  }
}

data class GatewayEvent(
  val type: GatewayEventType,
  val wireType: String,
  val runtimeSessionId: String?,
  val payload: JsonObject,
)

sealed interface JsonRpcFrame {
  data class Success(val id: String, val result: JsonObject) : JsonRpcFrame
  data class Failure(val id: String?, val code: Int, val message: String) : JsonRpcFrame
  data class Event(val event: GatewayEvent) : JsonRpcFrame
}

object JsonRpcCodec {
  fun encodeRequest(id: String, method: String, params: JsonObject = buildJsonObject {}): String {
    if (id.isBlank() || method.isBlank()) throw ProtocolException("RPC id and method are required")
    val frame = buildJsonObject {
      put("jsonrpc", "2.0")
      put("id", id)
      put("method", method)
      put("params", params)
    }
    return protocolJson.encodeToString(JsonElement.serializer(), frame)
  }

  fun decode(raw: String): JsonRpcFrame {
    val root = runCatching { protocolJson.parseToJsonElement(raw) }
      .getOrElse { throw ProtocolException("Invalid JSON-RPC frame") }
      as? JsonObject ?: throw ProtocolException("JSON-RPC frame must be an object")
    if (root.string("jsonrpc") != "2.0") throw ProtocolException("Unsupported JSON-RPC version")

    if (root.string("method") == "event") {
      val params = root["params"] as? JsonObject
        ?: throw ProtocolException("Event frame is missing params")
      val wireType = params.string("type")
        ?: throw ProtocolException("Event frame is missing type")
      return JsonRpcFrame.Event(
        GatewayEvent(
          type = GatewayEventType.fromWire(wireType),
          wireType = wireType,
          runtimeSessionId = params.string("session_id"),
          payload = params["payload"] as? JsonObject ?: buildJsonObject {},
        ),
      )
    }

    val id = root.string("id") ?: throw ProtocolException("Response frame is missing id")
    val error = root["error"] as? JsonObject
    if (error != null) {
      return JsonRpcFrame.Failure(
        id = id,
        code = error.int("code") ?: -1,
        message = error.string("message") ?: "Gateway request failed",
      )
    }
    val result = root["result"] as? JsonObject
      ?: throw ProtocolException("Response frame is missing result")
    return JsonRpcFrame.Success(id, result)
  }
}

data class SessionSummary(
  val storedId: String,
  val title: String,
  val preview: String,
  val startedAt: Long,
  val lastActive: Long,
  val messageCount: Int,
  val source: String,
  val archived: Boolean = false,
) {
  val displayTitle: String get() = title.ifBlank { preview.ifBlank { "New conversation" } }
}

enum class MessageRole {
  USER,
  ASSISTANT,
  TOOL,
  SYSTEM,
  UNKNOWN;

  companion object {
    fun fromWire(value: String): MessageRole = when (value.lowercase()) {
      "user" -> USER
      "assistant" -> ASSISTANT
      "tool" -> TOOL
      "system" -> SYSTEM
      else -> UNKNOWN
    }
  }
}

data class ChatMessageRecord(
  val rowId: Long?,
  val role: MessageRole,
  val content: String,
  val timestamp: Long?,
  val pending: Boolean = false,
  val error: String? = null,
)

data class ActiveSession(
  val runtimeId: String,
  val storedId: String,
  val messages: List<ChatMessageRecord>,
  val running: Boolean,
  val status: String,
  val model: String,
  val provider: String,
  val reasoningEffort: String,
)

object GatewayProtocol {
  fun parseSessionList(result: JsonObject): List<SessionSummary> =
    (result["sessions"] as? JsonArray).orEmpty().mapNotNull { element ->
      val value = element as? JsonObject ?: return@mapNotNull null
      val id = value.string("id")?.takeIf(String::isNotBlank) ?: return@mapNotNull null
      SessionSummary(
        storedId = id,
        title = value.string("title").orEmpty(),
        preview = value.string("preview").orEmpty(),
        startedAt = value.long("started_at") ?: 0L,
        lastActive = value.long("last_active") ?: value.long("started_at") ?: 0L,
        messageCount = value.int("message_count") ?: 0,
        source = value.string("source").orEmpty(),
        archived = value.boolean("archived") ?: false,
      )
    }

  fun parseActiveSession(result: JsonObject): ActiveSession {
    val runtimeId = result.string("session_id")
      ?: throw ProtocolException("Session response is missing runtime session id")
    val info = result["info"] as? JsonObject ?: buildJsonObject {}
    val storedId = result.string("resumed")
      ?: result.string("stored_session_id")
      ?: info.string("stored_session_id")
      ?: throw ProtocolException("Session response is missing durable session id")
    return ActiveSession(
      runtimeId = runtimeId,
      storedId = storedId,
      messages = parseMessages(result["messages"] as? JsonArray),
      running = result.boolean("running") ?: info.boolean("running") ?: false,
      status = result.string("status") ?: if (result.boolean("running") == true) "streaming" else "idle",
      model = info.string("model").orEmpty(),
      provider = info.string("provider").orEmpty(),
      reasoningEffort = info.string("reasoning_effort").orEmpty(),
    )
  }

  fun parseMessages(elements: JsonArray?): List<ChatMessageRecord> = elements.orEmpty().mapNotNull { element ->
    val value = element as? JsonObject ?: return@mapNotNull null
    val content = value.string("content") ?: value.string("text").orEmpty()
    ChatMessageRecord(
      rowId = value.long("id"),
      role = MessageRole.fromWire(value.string("role").orEmpty()),
      content = content,
      timestamp = value.long("timestamp"),
      pending = value.boolean("pending") ?: false,
      error = value.string("error"),
    )
  }
}

internal fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)?.contentOrNull()
internal fun JsonObject.long(key: String): Long? = string(key)?.toLongOrNull()
internal fun JsonObject.int(key: String): Int? = string(key)?.toIntOrNull()
internal fun JsonObject.boolean(key: String): Boolean? = string(key)?.toBooleanStrictOrNull()
internal fun JsonObject.objectValue(key: String): JsonObject = this[key] as? JsonObject
  ?: throw ProtocolException("JSON object field '$key' is missing")
internal fun JsonPrimitive.contentOrNull(): String? = content

File: apps/mobile/ios/HermesMobile/Core/GatewayProtocol.swift
import Foundation

public enum GatewayProtocolError: Error, Equatable, Sendable {
    case invalidResult
    case missingRuntimeSessionID
    case missingStoredSessionID
}

public enum ModelSwitchResult: Equatable, Sendable {
    case applied
    case confirmationRequired(message: String)
}

public struct PendingModelConfirmation: Equatable, Sendable {
    public let option: ModelOption
    public let message: String
    public let runtimeID: String

    public init(option: ModelOption, message: String, runtimeID: String) {
        self.option = option
        self.message = message
        self.runtimeID = runtimeID
    }
}

public enum MessageRole: String, Codable, Equatable, Hashable, Sendable {
    case user
    case assistant
    case tool
    case system
    case unknown

    public init(wireValue: String?) {
        switch wireValue?.lowercased() {
        case "user": self = .user
        case "assistant": self = .assistant
        case "tool": self = .tool
        case "system": self = .system
        default: self = .unknown
        }
    }
}

public struct SessionSummary: Identifiable, Codable, Equatable, Hashable, Sendable {
    public let storedID: String
    public var title: String
    public var preview: String
    public var startedAt: Double
    public var lastActive: Double
    public var messageCount: Int
    public var source: String
    public var archived: Bool
    public var pinned: Bool

    public var id: String { storedID }
    public var displayTitle: String {
        let candidate = title.trimmingCharacters(in: .whitespacesAndNewlines)
        if !candidate.isEmpty { return candidate }
        let fallback = preview.trimmingCharacters(in: .whitespacesAndNewlines)
        return fallback.isEmpty ? String(localized: "session.new") : fallback
    }

    public init(
        storedID: String,
        title: String = "",
        preview: String = "",
        startedAt: Double = 0,
        lastActive: Double = 0,
        messageCount: Int = 0,
        source: String = "",
        archived: Bool = false,
        pinned: Bool = false
    ) {
        self.storedID = storedID
        self.title = title
        self.preview = preview
        self.startedAt = startedAt
        self.lastActive = lastActive
        self.messageCount = messageCount
        self.source = source
        self.archived = archived
        self.pinned = pinned
    }
}

public struct ModelOption: Identifiable, Equatable, Hashable, Sendable {
    public let providerID: String
    public let providerName: String
    public let modelID: String
    public let supportsFast: Bool
    public let supportsReasoning: Bool

    public var id: String { modelID }

    public init(
        providerID: String,
        providerName: String,
        modelID: String,
        supportsFast: Bool = false,
        supportsReasoning: Bool = false
    ) {
        self.providerID = providerID
        self.providerName = providerName
        self.modelID = modelID
        self.supportsFast = supportsFast
        self.supportsReasoning = supportsReasoning
    }
}

public struct ModelProviderOption: Identifiable, Equatable, Sendable {
    public let id: String
    public let name: String
    public let models: [ModelOption]
    public let apiURL: String?

    public init(id: String, name: String, models: [ModelOption], apiURL: String? = nil) {
        self.id = id
        self.name = name
        self.models = models
        self.apiURL = apiURL
    }
}

public struct ModelCatalog: Equatable, Sendable {
    public let currentModel: String
    public let currentProvider: String
    public let providers: [ModelProviderOption]

    public init(
        currentModel: String = "",
        currentProvider: String = "",
        providers: [ModelProviderOption] = []
    ) {
        self.currentModel = currentModel
        self.currentProvider = currentProvider
        self.providers = providers
    }
}

public struct ChatMessageRecord: Identifiable, Codable, Equatable, Sendable {
    public let rowID: Int64?
    public let role: MessageRole
    public let text: String
    public let timestamp: Double?
    public let pending: Bool
    public let error: String?
    public let reasoning: String

    public var id: String {
        if let rowID { return "row-\(rowID)" }
        return "message-\(text.hashValue)"
    }

    public init(
        rowID: Int64? = nil,
        role: MessageRole,
        text: String,
        timestamp: Double? = nil,
        pending: Bool = false,
        error: String? = nil,
        reasoning: String = ""
    ) {
        self.rowID = rowID
        self.role = role
        self.text = text
        self.timestamp = timestamp
        self.pending = pending
        self.error = error
        self.reasoning = reasoning
    }
}

public struct ActiveSession: Equatable, Sendable {
    public let runtimeID: String
    public let storedID: String
    public let messages: [ChatMessageRecord]
    public let running: Bool
    public let status: String
    public let model: String
    public let provider: String
    public let reasoningEffort: String

    public init(
        runtimeID: String,
        storedID: String,
        messages: [ChatMessageRecord] = [],
        running: Bool = false,
        status: String = "idle",
        model: String = "",
        provider: String = "",
        reasoningEffort: String = ""
    ) {
        self.runtimeID = runtimeID
        self.storedID = storedID
        self.messages = messages
        self.running = running
        self.status = status
        self.model = model
        self.provider = provider
        self.reasoningEffort = reasoningEffort
    }
}

public enum GatewayProtocol {
    public static func parseSessionList(result: JSONValue) -> [SessionSummary] {
        let values = result.object?["sessions"]?.array
            ?? result.object?["data"]?.array
            ?? []
        return values.compactMap { value in
            guard let object = value.object,
                  let storedID = object["id"]?.stringValue,
                  !storedID.isEmpty else { return nil }
            return SessionSummary(
                storedID: storedID,
                title: object["title"]?.stringValue ?? "",
                preview: object["preview"]?.stringValue ?? "",
                startedAt: object["started_at"]?.numberValue ?? 0,
                lastActive: object["last_active"]?.numberValue ?? object["started_at"]?.numberValue ?? 0,
                messageCount: Int(object["message_count"]?.numberValue ?? 0),
                source: object["source"]?.stringValue ?? "",
                archived: object["archived"]?.boolValue ?? false,
                pinned: object["pinned"]?.boolValue ?? false
            )
        }
    }

    public static func parseActiveSession(result: JSONValue) throws -> ActiveSession {
        guard let object = result.object,
              let runtimeID = object["session_id"]?.stringValue,
              !runtimeID.isEmpty else { throw GatewayProtocolError.missingRuntimeSessionID }
        let info = object["info"]?.object ?? [:]
        let storedID = object["resumed"]?.stringValue
            ?? object["stored_session_id"]?.stringValue
            ?? info["stored_session_id"]?.stringValue
        guard let storedID, !storedID.isEmpty else { throw GatewayProtocolError.missingStoredSessionID }
        let running = object["running"]?.boolValue ?? info["running"]?.boolValue ?? false
        return ActiveSession(
            runtimeID: runtimeID,
            storedID: storedID,
            messages: parseMessages(object["messages"]),
            running: running,
            status: object["status"]?.stringValue ?? (running ? "streaming" : "idle"),
            model: info["model"]?.stringValue ?? "",
            provider: info["provider"]?.stringValue ?? "",
            reasoningEffort: info["reasoning_effort"]?.stringValue ?? ""
        )
    }

    public static func parseModelSwitch(result: JSONValue) -> ModelSwitchResult {
        let root = result.object ?? [:]
        if root["confirm_required"]?.boolValue == true {
            let message = root["confirm_message"]?.stringValue
                ?.trimmingCharacters(in: .whitespacesAndNewlines)
            return .confirmationRequired(message: message?.isEmpty == false ? message! : String(localized: "model.confirm.default"))
        }
        return .applied
    }

    public static func parseModelOptions(result: JSONValue) -> ModelCatalog {
        let root = result.object ?? [:]
        let providers = (root["providers"]?.array ?? []).compactMap { value -> ModelProviderOption? in
            guard let object = value.object,
                  object["authenticated"]?.boolValue == true,
                  let providerID = object["slug"]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !providerID.isEmpty else { return nil }
            let candidateName = object["name"]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let providerName = candidateName.isEmpty ? providerID : candidateName
            let capabilities = object["capabilities"]?.object ?? [:]
            var seen = Set<String>()
            let models = (object["models"]?.array ?? []).compactMap { item -> ModelOption? in
                guard let modelID = item.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines),
                      !modelID.isEmpty,
                      seen.insert(modelID).inserted else { return nil }
                let modelCapabilities = capabilities[modelID]?.object ?? [:]
                return ModelOption(
                    providerID: providerID,
                    providerName: providerName,
                    modelID: modelID,
                    supportsFast: modelCapabilities["fast"]?.boolValue ?? false,
                    supportsReasoning: modelCapabilities["reasoning"]?.boolValue ?? false
                )
            }
            return ModelProviderOption(
                id: providerID,
                name: providerName,
                models: models,
                apiURL: object["api_url"]?.stringValue
            )
        }
        return ModelCatalog(
            currentModel: root["model"]?.stringValue ?? "",
            currentProvider: root["provider"]?.stringValue ?? "",
            providers: providers
        )
    }

    public static func parseMessages(_ value: JSONValue?) -> [ChatMessageRecord] {
        (value?.array ?? []).compactMap { item in
            guard let object = item.object else { return nil }
            let text = object["content"]?.stringValue ?? object["text"]?.stringValue ?? ""
            return ChatMessageRecord(
                rowID: object["id"]?.numberValue.map { Int64($0) },
                role: MessageRole(wireValue: object["role"]?.stringValue),
                text: text,
                timestamp: object["timestamp"]?.numberValue,
                pending: object["pending"]?.boolValue ?? false,
                error: object["error"]?.stringValue,
                reasoning: object["reasoning"]?.stringValue ?? object["reasoning_content"]?.stringValue ?? ""
            )
        }
    }
}